### PR TITLE
Support for Kubernetes v1.22

### PR DESCRIPTION
**Pull request by timuthy**
**How to categorize this PR?**

/area open-source usability
/kind enhancement
/priority 3
/exp intermediate
/topology garden seed shoot

**What this PR does / why we need it**:
Support for Kubernetes v1.22

**Which issue(s) this PR fixes**:
Parts of #4363 

**Special notes for your reviewer**:
* I didn't vendor new `k8s.io/*` Golang sources yet as we usually do this together with a new `sigs.k8s.io/controller-runtime` upgrade.
* ⚠️ This PR only contains the Gardener part for supporting 1.22 - there will/must be follow-up PRs for each extension repository where individual support gets added.
* Code changes have been tested with adjusted versions of [provider-gcp](https://github.com/gardener/gardener-extension-provider-gcp/tree/kubernetes-1.22) -> [gardener/gardener-extension-provider-gcp#314](https://github.com/gardener/gardener-extension-provider-gcp/pull/314)and [provider-aws](https://github.com/gardener/gardener-extension-provider-aws/tree/kubernetes-v1.22) ->  [gardener/gardener-extension-provider-aws#396](https://github.com/gardener/gardener-extension-provider-aws/pull/396).
** Create/Delete a new shoot cluster with v1.22.1
** Upgrade an existing cluster to v1.22.1
** Register a v1.22.1 as a seed cluster
* I upgrade any affected API version that it removed with Kubernetes v1.22 ([ref](https://kubernetes.io/docs/reference/using-api/deprecation-guide/#v1-22)) and let it be deployed on the lowest supported Kubernetes version, e.g. `CustomResourceDefinition apiextensions.k8s.io/v1` is now used for every version >= v1.16 (instead of just using for for >= v1.22). This ensures that we can clean-up code and manifests as soon as we have dropped support for 1.15. The same applies for other APIs and versions.
* ⚠️ Garden clusters with version v1.22 are not supported yet. As a follow-up item, we need to add a proper handling for the removed `CSR certificates.k8s.io/v1beta1` API to the Gardener-Controller-Manager and Gardenlet.

**Release note**:

```feature operator
Gardener can now support shoot and seed clusters with Kubernetes version 1.22. In order to allow creation/update of 1.22 clusters you will have to update the version of your provider extension(s) to a version that supports 1.22 as well. Please consult the respective releases and notes in the provider extension's repository.
```
```feature developer
Gardener can now support shoot clusters with Kubernetes version 1.22. Extension developers have to prepare individual extensions as well to work with 1.22. Please take extra care to check the built-in APIs that have been removed with Kubernetes v1.22 ([ref](https://kubernetes.io/docs/reference/using-api/deprecation-guide/#v1-22)). Those need to be replaced by a corresponding newer API version.
```

### CODE REVIEW

**Comment by timuthy** [comment]
FYI: I'm working on fixing and enhancing the failing test to also cover a additional code branch.

File: README.md [hunk 1]
```diff
@@ -38,15 +38,15 @@ Gardener takes part in the [Certified Kubernetes Conformance Program](https://ww
...
+| **Packet** | N/A | N/A | N/A | N/A | N/A | N/A | N/A | N/A |
```
Comment (rfranzke) [inline README.md]:
```suggestion
| **Equinix Metal** | N/A | N/A | N/A | N/A | N/A | N/A | N/A | N/A |
```

File: pkg/utils/kubernetes/tls_cipher_suites.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubernetes
+
+import (
+	"github.com/Masterminds/semver"
+
```
Comment (rfranzke) [inline pkg/utils/kubernetes/tls_cipher_suites.go:19]:
```suggestion
```

File: pkg/utils/version/version.go [hunk 3]
```diff
@@ -18,9 +18,63 @@ import (
...
+	// ConstraintK8sEqual120 is a version constraint for versions == 1.20.
+	ConstraintK8sEqual120 *semver.Constraints
+	// ConstraintK8sLessEqual121 is a version constraint for versions <= 1.21.
+	ConstraintK8sLessEqual121 *semver.Constraints
+)
+
+func init() {
+	var err error
+
+	ConstraintK8sLessEqual115, err = semver.NewConstraint("<= 1.15.x")
```
Comment (rfranzke) [inline pkg/utils/version/version.go:59]: Why `.x`?

File: pkg/utils/version/version.go [hunk 4]
```diff
@@ -18,9 +18,63 @@ import (
 	"fmt"
 	"strings"
 
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+
 	"github.com/Masterminds/semver"
 )
 
+var (
```
Comment (rfranzke) [inline pkg/utils/version/version.go:25]: Great idea!

File: pkg/utils/version/version.go [hunk 5]
```diff
@@ -18,9 +18,63 @@ import (
 	"fmt"
 	"strings"
 
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+
```
Comment (rfranzke) [inline pkg/utils/version/version.go]:
```suggestion
```

File: extensions/pkg/controller/controlplane/genericactuator/actuator_test.go [hunk 6]
```diff
@@ -30,16 +31,18 @@ import (
 	mockclient "github.com/gardener/gardener/pkg/mock/controller-runtime/client"
 	"github.com/gardener/gardener/pkg/utils/chart"
 	mockchartutil "github.com/gardener/gardener/pkg/utils/chart/mocks"
-	"github.com/gardener/gardener/pkg/utils/imagevector"
 	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
 	mocksecretsutil "github.com/gardener/gardener/pkg/utils/secrets/mock"
+	"k8s.io/apimachinery/pkg/runtime"
```
Comment (rfranzke) [inline extensions/pkg/controller/controlplane/genericactuator/actuator_test.go]:
```suggestion

	"k8s.io/apimachinery/pkg/runtime"
````

File: pkg/operation/botanist/component/kubescheduler/kube_scheduler_test.go [hunk 7]
```diff
@@ -445,6 +445,8 @@ var _ = Describe("KubeScheduler", func() {
 				Entry("kubernetes 1.18 w/ full config", "1.18.8", configFull),
 				Entry("kubernetes 1.19 w/o config", "1.19.9", configEmpty),
 				Entry("kubernetes 1.19 w/ full config", "1.19.9", configFull),
+				Entry("kubernetes 1.22 w/o config", "1.22.1", configEmpty),
+				Entry("kubernetes 1.22 w/ full config", "1.22.1", configFull),
```
Comment (rfranzke) [inline pkg/operation/botanist/component/kubescheduler/kube_scheduler_test.go:453]: What about 1.20/1.21 tests here?

**Review by rfranzke — COMMENTED** [review]
Great work, highly appreciated, thanks @timuthy!

File: pkg/utils/version/version.go (hunk 3 above)
Comment (timuthy) [reply pkg/utils/version/version.go:59]: See https://play.golang.org/p/uXF9vOkPtT9

**Comment by timuthy** [comment]
Thanks for your review @rfranzke. I addressed all your comments and fixed a VPA issue I noticed while bootstrapping a new seed cluster.

**Review by rfranzke — APPROVED** [review]
/lgtm

**Pull request merged**